Improve metrics performance and reliability

santigimeno opened the pull request:
```
test: refactor OTLP metric validators

Move common validator logic for gRPC to test/common/nsolid-grpc-agent/validators.js.
Add a new validator for OTLP Metrics: [checkOTLPMetricsData()](cci:1://file:///home/sgimeno/nodesource/nsolid/test/common/nsolid-grpc-agent/validators.js:157:0-267:1).
Add also a helper method [hasThreadAttributes()](cci:1://file:///home/sgimeno/nodesource/nsolid/test/common/nsolid-grpc-agent/validators.js:94:0-98:1) to check whether a metric datapoint is a thread or a process.
Refactor `test/agents/test-grpc-metrics.mjs` to use these new helpers.
```
```
agents: add MetricDataBatch for OTLP metrics

Introduce MetricDataBatch class to consolidate metric collection before export, reducing per-metric overhead and improving batch processing efficiency.
Update OTLP/grpc agents to accumulate metrics in batches and dump them into ScopeMetrics during export.
This change reduces the number of individual metric exports and improves performance for high-frequency metric generation.

Changes:
- Introduce MetricDataBatch class for metric accumulation.
- Update OTLP agents to use batched metric collection.
- Modify test expectations to handle batched datapoints.
```
```
agents: generalize DelegateAsyncExport

Refactor DelegateAsyncExport to accept generic stub, event, and response types as template parameters, enabling reuse across different gRPC services beyond the original NSolidService. This generalization prepares the codebase for supporting the OTLP MetricsService exporter while maintaining backward compatibility with existing event exports.

Key changes:
- Template DelegateAsyncExport with <Stub, EventT, ResponseT> params.
- Update GrpcAsyncCallData to use generic ResponseType.
- Change callback signature from bool return to void return.
- Update all existing DelegateAsyncExport call sites.
- Remove unnecessary “return true” statements from callbacks.
```
```
core: add RingBuffer resizing with coverage

Allow RingBuffer to resize at runtime by moving the newest entries into a freshly allocated buffer and resetting head/tail bookkeeping.
Require strictly positive capacities via CHECK_GT and drop the const on capacity_ so it can reflect the current size.
Add gtest coverage for all resizing scenarios (no-op, grow empty/full, shrink with and without wraparound, death test for resize(0)) to prevent regressions.
```
```
agents: add async OTLP GrpcMetricsExporter

Introduce GrpcMetricsExporter that uses DelegateAsyncExport + AsyncTSQueue to flush MetricDataBatch instances off the main thread and retry transient gRPC failures with backpressure.
Add NSOLID_METRICS_BUFFER_SIZE option (plumbed through lib/nsolid.js, reconfigure proto/pb.cc/pb.h, and node.gyp) plus new test-grpc-metrics-buffer-size.mjs and retry coverage. It defaults to 100 elements, following the ZmqAgent behavior.
Wire GrpcAgent config/reconfigure paths to construct the exporter once and call resize_buffer on buffer-size changes, ensuring the new queue depth propagates without losing metrics.
```
```
lib: validate metrics batch/buffer config inputs

Require positive finite numbers when metricsBatchSize/metricsBufferSize are set via nsolid.start() or live reconfigure so odd JS values (true, objects, NaN, etc.) are ignored.
Add test-nsolid-start-metrics-buffer.js to cover these cases.
```

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added metrics batching and buffering capabilities with configurable batch size and buffer size parameters for improved metrics export efficiency.
  * Introduced dynamic metrics reconfiguration allowing batch size and buffer size adjustments at runtime without restart.
  * Added metrics pause and resume functionality for better control over metric collection during operations.

* **Improvements**
  * Enhanced metrics export resilience with improved retry logic and handling for interrupted connections.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

santigimeno committed:
agents: add MetricDataBatch for OTLP metrics

Introduce MetricDataBatch class to consolidate metric collection before
export, reducing per-metric overhead and improving batch processing
efficiency. Update OTLP/grpc agents to accumulate metrics in batches
and dump them into ScopeMetrics during export.

This change reduces the number of individual metric exports and
improves performance for high-frequency metric generation.

Changes:
- Introduce MetricDataBatch class for metric accumulation.
- Update OTLP agents to use batched metric collection.
- Modify test expectations to handle batched datapoints.

santigimeno committed:
agents: generalize DelegateAsyncExport

Refactor DelegateAsyncExport to accept generic stub, event, and response
types as template parameters, enabling reuse across different gRPC
services beyond the original NSolidService. This generalization prepares
the codebase for supporting the OTLP MetricsService exporter while
maintaining backward compatibility with existing event exports.

Key changes:
- Template DelegateAsyncExport with <Stub, EventT, ResponseT> params.
- Update GrpcAsyncCallData to use generic ResponseType.
- Change callback signature from bool return to void return.
- Update all existing DelegateAsyncExport call sites.
- Remove unnecessary return true statements from callbacks.

santigimeno committed:
agents: add async OTLP GrpcMetricsExporter

Introduce GrpcMetricsExporter that uses DelegateAsyncExport +
AsyncTSQueue to flush MetricDataBatch instances off the main thread and
retry transient gRPC failures with backpressure.
Add NSOLID_METRICS_BUFFER_SIZE option (plumbed through lib/nsolid.js,
reconfigure proto/pb.cc/pb.h, and node.gyp) plus new
test-grpc-metrics-buffer-size.mjs and retry coverage. It defaults to 100
elements, following the ZmqAgent behavior.
Wire GrpcAgent config/reconfigure paths to construct the exporter once and
call resize_buffer on buffer-size changes, ensuring the new queue depth
propagates without losing metrics.

santigimeno committed:
core: add RingBuffer resizing with coverage

Allow RingBuffer to resize at runtime by moving the newest entries into
a freshly allocated buffer and resetting head/tail bookkeeping.
Require strictly positive capacities via CHECK_GT and drop the const on
capacity_ so it can reflect the current size.
Add gtest coverage for all resizing scenarios (no-op, grow empty/full,
shrink with and without wraparound, death test for resize(0)) to prevent
regressions.

santigimeno committed:
test: refactor OTLP metric validators

Move common validator logic for gRPC to
test/common/nsolid-grpc-agent/validators.js.
Add a new validator for OTLP Metrics: `checkOTLPMetricsData()`.
Add also a helper method `hasThreadAttributes()` to check whether a
metric datapoint is a thread or a process.
Refactor `test/agents/test-grpc-metrics.mjs` to use these new helpers.

santigimeno committed:
lib: validate metrics batch/buffer config inputs

Add normalizePositiveFinite() helper that parses a raw value (env
string, pkg value, or JS API input) to a positive finite number,
returning a fallback when invalid.

Refactor both initializeConfig() and updateConfig() to use the shared
helper for metricsBatchSize and metricsBufferSize. This fixes
initializeConfig() which previously used the +env || pkg || default
pattern and would accept Infinity or negative numbers from environment
variables or package.json.

Expand test-nsolid-start-metrics-buffer.js to cover Infinity,
-Infinity, and string variants of invalid values.

Add test-nsolid-config-metrics-env.js to verify that invalid env
values (Infinity, negative, zero, non-numeric) correctly fall back
to defaults during initialization.

santigimeno commented:
@coderabbitai review

RafaelGSS commented on lib/nsolid.js:
```diff
@@ -706,6 +710,14 @@ function updateConfig(config = {}) {
+    } else if (key === 'metricsBatchSize' || key === 'metricsBufferSize') {
+      const raw = config[key];
+      if (typeof raw === 'number' || (typeof raw === 'string' && raw.trim() !== '')) {
+        const value = Number(raw);
```
Should we check if the value is `NaN` after this conversion?

santigimeno replied:
Yes, it should be solved now

santigimeno committed:
agents: implement custom OTLP metrics serializer

Introduce custom OTLP metrics serialization in gRPC agent,
replacing SDK with BatchedMetricData and direct protobuf
generation. Provides per-data-point timestamps, custom batching,
and improved performance for high-frequency metrics.

Add debug logging, fix thread ID type mismatches, implement
robust batch processing for partial metrics, and update test
expectations to resolve inconsistent processing and hangs.

RafaelGSS reviewed (approved):
RSLGTM

santigimeno committed:
agents: implement gRPC retry policies

Implement client-side retry policies for gRPC connections to improve
resilience against transient network failures. The retry policy
automatically retries failed requests on UNAVAILABLE status with
exponential backoff.

Retry policy configuration:
- Max attempts: 5 retries
- Initial backoff: 500ms
- Max backoff: 5 seconds
- Backoff multiplier: 2.0x
- Retryable status: UNAVAILABLE

The retry policy is configured via gRPC service config JSON and applies
to all RPC methods. This ensures robust connectivity to the N|Solid
Console even during temporary network issues or server restarts.

santigimeno committed:
test: add gRPC retry policy tests

Add comprehensive test suite to verify gRPC retry policy behavior.
Extend test infrastructure with fault injection capabilities to simulate
transient failures and network issues.

Test infrastructure enhancements:
- Inject transient failures with configurable status codes and counts
- Inject delays to simulate network latency
- Track retry attempts via grpc-previous-rpc-attempts metadata
- Clear all injected faults between tests

Test coverage:
- Successful retry after single transient UNAVAILABLE failure
- Failure after exhausting max retries (5 attempts) on persistent errors
- Verification of retry attempt counts via metadata headers
- Multiple service types (ExportInfo, ExportMetricsCmd, ExportSpans)

The fault injection system validates that the gRPC client properly
retries on transient failures and respects retry limits.

santigimeno commented on agents/grpc/src/grpc_agent.cc:
```diff
@@ -920,20 +916,22 @@ void GrpcAgent::env_deletion_cb_(SharedEnvInst envinst,
-    std::vector<ScopeMetrics>{{otlp::GetScope(), metrics}};
-  auto result = agent->metrics_exporter_->Export(data);
-  Debug("# ThreadMetrics Exported. Result: %d\n", static_cast<int>(result));
+  if (agent->thr_metrics_batch_.ShouldFlush()) {
+    auto metric_data = agent->thr_metrics_batch_.DumpMetricsAndReset();
+    if (!metric_data.empty()) {
+      ResourceMetrics data;
+      data.resource_ = otlp::GetResource();
+      data.scope_metric_data_ =
+        std::vector<ScopeMetrics>{{ otlp::GetScope(), std::move(metric_data) }};
+      agent->metrics_exporter_->enqueue(std::move(data));
+    }
+  }
```
Both `do_stop` and `metrics_msg_cb_` run in the `GrpcAgent` `thread_`, so I don't think this can happen. Can you deeply analyze this to confirm?

santigimeno commented on lib/nsolid.js:
```diff
@@ -847,6 +859,18 @@ function initializeConfig(nsolidConfig) {
     pkgConfig.nsolid.interval ||
     DEFAULT_INTERVAL;
 
+  // Metrics batch size
+  nsolidConfig.metricsBatchSize =
+    +process.env.NSOLID_METRICS_BATCH_SIZE ||
+    pkgConfig.nsolid.metricsBatchSize ||
+    DEFAULT_METRICS_BATCH_SIZE;
+
+  // Metrics buffer size
+  nsolidConfig.metricsBufferSize =
+    +process.env.NSOLID_METRICS_BUFFER_SIZE ||
+    pkgConfig.nsolid.metricsBufferSize ||
+    DEFAULT_METRICS_BUFFER_SIZE;
+
```
I think this should be solved by last push but please, take a look,
